### PR TITLE
Rotation support

### PR DESCRIPTION
**Pull request by dai-shi**
This is to support rotation. #15

- [x] basic feature
- [x] fix error in arrow/line bounds
- [x] fix error with rotated lines/arrows when angle > Math.PI
- [x] fix resizing with "w*" or "n*" leads elements moving.
- [x] png export
- [x] svg export

----

Some small remaining issues not addressed in this PR:

- rotate cursor is "ew".
- resize cursors don't change with rotation.
- `resize(Test|Handles)` should be renamed because they are used for "rotation" now.

### CODE REVIEW

**Comment by dai-shi** [comment]
My initial focus is to render a rectangle with a fixed angle (= `Math.PI / 4`).

**Comment by dai-shi** [comment]
Thanks to @dwelle help, I've got this.

rect1

**Comment by dwelle** [comment]
Hi. Thanks for working on this.

Some answers to your questions from twitter https://mobile.twitter.com/dai_shi/status/1243925468375429120:

> Resizing is not intuitive. (How to fix rotated edges...)

Not sure what you mean by rotated edges?

If you meant why the width/height/position changes in weird ways during resize, it's likely due to normalization and stuff we do during resize in:

https://github.com/excalidraw/excalidraw/blob/aa54364bd687d6149ccc25098e465c835a07ca3e/src/components/App.tsx#L1935-L2110

> Handler mouse pointers are not rotated. (Can be create arbitrary pointers with various angles?)

Native browser cursors can't be rotated AFAIK, so the best we can do is to hide it and replace it with a custom cursor: https://css-tricks.com/can-you-rotate-the-cursor-in-css/

**Comment by dai-shi** [comment]
Resizing is mostly done. I wonder if I should spend time on custom cursors or go on rotating rectangles.

rect3

**Comment by vjeux** [comment]
I feel like the cursors are fine right now. I wouldn't spend too much time on that.

**Comment by dai-shi** [comment]
Found a bug when `angle > Math.PI / 2`. 🤔

**Comment by dai-shi** [comment]
naive rotation!

rect4

**Comment by lipis** [comment]
Tests are not running on drafts.. making it a PR.

**Comment by dai-shi** [comment]
So much for my experiments this weekend. Remaining:

- Resize position is not very correct for "N", "W", and "NW".
- Rotate handler is too naive.
- Apply for non-rectangle shapes.

Anybody interested in investigating?

**Comment by j-f1** [comment]
Interesting choice for the rotation handles! How do we feel about using this instead of the more usual method of adding an additional handle?

**Comment by dai-shi** [comment]
I just realized that I don't need to rotate on each elements but rotate before `drawImag`.

rect5

**Comment by lipis** [comment]
Not obvious and what about mobile?

**Comment by dai-shi** [comment]
It doesn't work on mobile yet. Rotation requires SHIFT at the moment. Any suggestion?

**Comment by lipis** [comment]
Something more classic? One idea is this:

Screenshot 2020-03-30 at 11 35 59 AM

**Comment by dai-shi** [comment]
Here you go!

rect6

**Comment by lipis** [comment]
Looking good! Will you tackle the text element in another PR?

**Comment by dai-shi** [comment]
@lipis Do you mean you will merge this PR before trying text elements? I'm OK with it.
Any idea how to pass the failing test? I'm using "rotation" handler as one of "resize" handlers, which may cause the test fails.

**Comment by lipis** [comment]
> @lipis Do you mean you will merge this PR before trying text elements? I'm OK with it.

It would be nice to get the text elements as well, but it's up to you.. I'm OK with having it done in another PR.

As for the tests, I'm not sure what's going on. Let's wait for someone to check it out :)

**Comment by dai-shi** [comment]
@lipis Yay!

rect7

**Comment by lipis** [comment]
What about save/load..?

**Comment by dai-shi** [comment]
I assume it works out of the box. Yeah, looks like working.
```json
{
  "type": "excalidraw",
  "version": 1,
  "source": "http://localhost:3000",
  "elements": [
    {
      "id": "Uh5zdFkKHGULieRRuZ7l5",
      "type": "rectangle",
      "x": 442.80859375,
      "y": 137.87109375,
      "width": 172.38671875,
      "height": 62.62890625,
      "angle": 0.5833040710692037,
      "strokeColor": "#000000",
      "backgroundColor": "#ced4da",
      "fillStyle": "hachure",
      "strokeWidth": 1,
      "roughness": 1,
      "opacity": 100,
      "seed": 671820596,
      "version": 111,
      "versionNonce": 1172941580,
      "isDeleted": false
    }
  ],
  "appState": {
    "viewBackgroundColor": "#ffffff"
  }
}
```

**Comment by dwelle** [comment]
Amazing progress!

One problem I noticed you may be already aware of:

Upon rotation (esp more than 180deg), resizing doesn't keep position correctly (the element floats). For arrows it's even weirder (2nd part of the gif):

excalidraw_rotate_resize_bug

**Comment by dai-shi** [comment]
Rotation locking with SHIFT. We need to changed the instruction.

rect8

**Comment by dai-shi** [comment]
@dwelle Yeah, I'm aware of that. (> resizing with "w*" or "n*" leads elements moving.)
I spent much time but didn't get what's happening. Note "s","e","se" handlers are OK.
Didn't know the wired behavior with arrows. I need someone to help on it.

**Comment by dwelle** [comment]
@dai-shi I can try to look at it, but don't have much time today/tomorrow.

**Comment by dai-shi** [comment]
It's working correctly now, for all features as far as I know.
Please find a bug!

> Didn't know the wired behavior with arrows.

This is fixed. That was because of my lack of the understanding about `canvasOffset`.

> Upon rotation (esp more than 180deg), resizing doesn't keep position correctly (the element floats).

This is not fixed. My workaround is noted as `FIXME`.

**Comment by lipis** [comment]
Yes, when resizing (from most of the points) it still moves the object around when rotated..

**Comment by dai-shi** [comment]
And, I found a bug. 😓 

rect9

**Comment by dai-shi** [comment]
@lipis Which shape is it?

**Comment by lipis** [comment]
All of them.. the most weird one is the arrow/line of course

**Comment by dai-shi** [comment]
@lipis I'm not sure if what you observe is a known one or a new one... 😓

**Comment by lipis** [comment]
New or old, would be nice to tackle it though..

**Comment by j-f1** [comment]
Why do the resize handles get so big? (Ideally, the handles would appear small but have large tap targets.

EVDD5989

File: src/locales/en.json [hunk 1]
```diff
@@ -97,7 +97,8 @@
   "hints": {
     "linearElement": "Click to start multiple points, drag for single line",
     "linearElementMulti": "Click on last point or press Escape or Enter to finish",
-    "resize": "You can constraint proportions by holding SHIFT while resizing"
+    "resize": "You can constraint proportions by holding SHIFT while resizing",
+    "rotate": "You can constraint angles by holding SHIFT while rotating"
```
Comment (jdestgermain) [inline src/locales/en.json]: I found it odd this was using the noun `constraint` rather than the verb `constrain`

**Comment by dai-shi** [comment]
> New or old, would be nice to tackle it though..

@lipis Hm, I probably don't understand your observation. Is it too hard for you to take a gif?

**Comment by dai-shi** [comment]
> Why do the resize handles get so big? (Ideally, the handles would appear small but have large tap targets.

@j-f1 I didn't know that it's intentional. The problem was that the _position_ of the rotation handler changes on mobile. Let me see what I can do.

File: src/locales/en.json (hunk 1 above)
Comment (dai-shi) [reply src/locales/en.json]: Oops, maybe we fix the "resize" message in another PR, and this PR will follow that? Would you or somebody be willing to create a PR?

**Comment by lipis** [comment]
Kapture 2020-03-31 at 17 20 35

File: src/locales/en.json (hunk 1 above)
Comment (lipis) [reply src/locales/en.json]:
```suggestion
    "resize": "You can constrain proportions by holding SHIFT while resizing",
    "rotate": "You can constrain angles by holding SHIFT while rotating"
```

**Comment by dai-shi** [comment]
> The opposite side/corner should remain still.

Thanks for the gif!

Yeah, this is something I spent the most of my development time so far without success. I will try it again, but maybe I'm falling into a hole, so I wish someone take a look with fresh eyes.
Just a note: "E", "N", "NE" handlers are working fine, all other 5 resize handlers are problematic.

**Comment by Jakst** [comment]
No need for the message about the SHIFT key modifier on mobile imo

**Comment by mochetts** [comment]
Hi @dai-shi I implemented svg shapes rotation for a similar tool last year. It was implemented using Vue.js but the concept should be the same. If you want, I can send you the code so that you can take a look at it.
rotation

**Comment by dai-shi** [comment]
@mochetts Looks cool. It renders with SVG in browser, right? Cursor pointers are nice.

**Comment by j-f1** [comment]
I’m not sure this is necessary or even a good idea but it might be useful to allow rotating on touchscreens by rotating with two fingers.

**Comment by vjeux** [comment]
In my experience it works well for large surfaces like wall screens or big tables. But for small surface like phones, your two fingers are basically on-top of the element and you can’t see anything.

But it’s worth playing around

**Comment by mochetts** [comment]
@dai-shi yes! It renders svg. 🙂

**Comment by pshihn** [comment]
The tricky part with touch rotate is if you want to support both pinch zoom and two-finger rotate, one should be able to do both before touchup happens. I have seen it work well for things like story creation in snpachat/instagram. (Tough for small shapes)

**Comment by dai-shi** [comment]
> Upon rotation (esp more than 180deg), resizing doesn't keep position correctly (the element floats).

@dwelle @lipis It's fixed now!!!

rect10

**Comment by dai-shi** [comment]
Still floating a bit with curved line/arrow...

**Comment by lipis** [comment]
Curved lines and arrows, were floating on production anyway... and they don't go to negatives there.

**Comment by lipis** [comment]
Screenshot 2020-04-01 at 3 19 06 PM

What happened to the circle for rotation..?

**Comment by lipis** [comment]
The rotation handle on text element is not there.

**Comment by dai-shi** [comment]
Sorry, broke with refactoring. Fixed with the latest commit.

And, all set!

**Comment by j-f1** [comment]
Do we want to rotate the handles? It looks like Figma does but Keynote doesn’t. Also I know I’m the person who asked for the rotation handle in the first place and still think it’s a good idea but I checked Keynote and Figma and they both support Cmd+drag on any of the handles to rotate, and Figma has an invisible area outside of each handle that you can drag to rotate. Maybe we should add some of these now or in the future?

**Comment by jdestgermain** [comment]
I like the rotation handle for mobile benefits.   I also wanted to add that the tool I previously used for drawing FBD's was Google Drawings and this is what theirs looks like:
image

Also note it shows you the rotation angle which is helpful

**Comment by lipis** [comment]
I would say in the future.. I like the current behaviour! We can improve as we go.

**Comment by dwelle** [comment]
Amazing work @dai-shi. I've pushed 2 minor commits:

- ensure we migrate old scene elements correctly by initializing `angle` to `0`
- rotate wysiwyg editor:

excalidraw_wysiwyg_rotate

I'm still reviewing the rest, but so far I don't see anything else.

File: src/renderer/renderElement.ts [hunk 2]
```diff
@@ -366,7 +371,9 @@ export function renderElementToSvg(
       }
       node.setAttribute(
         "transform",
-        `translate(${offsetX || 0} ${offsetY || 0})`,
+        `translate(${offsetX || 0} ${
+          offsetY || 0
+        }) rotate(${degree} ${cx} ${cy})`,
```
Comment (dwelle) [inline src/renderer/renderElement.ts:376]: AFAIK the `cx` `cy` args do nothing. `rotate()` takes only a single arg. https://developer.mozilla.org/en-US/docs/Web/CSS/transform-function/rotate

Comment (dwelle) [reply src/renderer/renderElement.ts:376]: Mhm.. it seems necessary for SVG. What is this sorcery.

Comment (dwelle) [reply src/renderer/renderElement.ts:376]: Ok, now I'm seeing it. It's not a CSS prop, but an SVG attribute, so nvm.

**Comment by lipis** [comment]
Doesn't lock on the left side :)

Kapture 2020-04-01 at 20 27 39

**Comment by dwelle** [comment]
I can repro. To elaborate: it skips the 270° almost all the time (though if you're really precise, it locks).

**Comment by dwelle** [comment]
Fixed via a hideous hack IMO.. anyone feel free to swap it with something better :).

**Comment by dai-shi** [comment]
Oh, nice catch. Let me try to find a better way.

8c38d91 Fixed with an improvement.

**Comment by lipis** [comment]
It feels like if we wait a bit longer we can shave a few more lines and optimize it more.. it will be easier to maintain later, while this topic is still hot.

P.S. @dai-shi you've done some amazing work!

**Comment by dai-shi** [comment]
There're some repeated patterns around `rotateHandlerCoords`, so there's probably room for improvement.

Other than that, it would require big refactoring for optimization, which is out of the scope of this PR, I think.

**Comment by lipis** [comment]
If you think of something go ahead.. we can merge this tomorrow. Let us know when you feel it's done.

**Comment by dai-shi** [comment]
Slightly improved, and the diff is cleaner.

@lipis I'm done. Please go head and merge it!

**Comment by dai-shi** [comment]
> Do we want to rotate the handles?

@j-f1 We don't need to. I think it looks nicer with rotated handles in our case. Note resizeTest uses unrotated handles anyway, so it's just how it looks.

rect11
rect12

**Comment by dwelle** [comment]
Stellar job @dai-shi! :tada: :rocket: :tada:

**Comment by Dentarthurdent42** [comment]
Is it possible to rotate the canvas? I’ve been trying on iPadOS because it’d make aligning things SO much less tedious.

**Pull request merged**